specutils/analysis/template_match.py

```diff
@@ -0,0 +1,105 @@
+from ..spectra.spectrum1d import Spectrum1D
+from ..spectra.spectrum_collection import SpectrumCollection
+from ..manipulation import FluxConservingResampler
+import numpy as np
+
+def _normalize(observed_spectrum, template_spectrum):
+    """
+    Calculate a scale factor to be applied to the template spectrum so the total flux in both spectra will be the same.
+
+    Parameters
+    ----------
+    observed_spectrum : :class:`~specutils.Spectrum1D`
+        the observed spectrum
+    template_spectrum : :class:`~specutils.Spectrum1D`
+        the template spectrum, which needs to be normalized in order to be compared with the observed_spectrum
+
+    Returns
+    -------
+    `float`
+        A float which will normalize the template spectrum's flux so that it can be compared to the observed spectrum
+    """
+    num = np.sum((observed_spectrum.flux*template_spectrum.flux)/(observed_spectrum.uncertainty.array**2))
+    denom = np.sum((template_spectrum.flux/observed_spectrum.uncertainty.array)**2)
+    return num/denom
+
+def _template_match(observed_spectrum, template_spectrum):
+    """
+    Resample the template spectrum to match the wavelength of the observed spectrum.
+    Then, calculate chi2 on the flux of the two spectra.
+
+    Parameters
+    ----------
+    observed_spectrum : :class:`~specutils.Spectrum1D`
+        the observed spectrum
+    template_spectrum : :class:`~specutils.Spectrum1D`
+        the template spectrum, which will be resampled to match the wavelength of the observed_spectrum
+
+    Returns
+    -------
+    normalized_template_spectrum : :class:`~specutils.Spectrum1D`
+        the template_spectrum that has been normalized
+    chi2 : `float`
+        the chi2 of the flux of the observed_spectrum and the flux of the normalized_template_spectrum
+    """
+    # Resample template
+    fluxc_resample = FluxConservingResampler()
+    template_obswavelength = fluxc_resample(template_spectrum, observed_spectrum.wavelength)
+
+    # Normalize spectra
+    normalization = _normalize(observed_spectrum, template_obswavelength)
+
+    # Numerator
+    num_right = normalization*template_obswavelength.flux
+    num = observed_spectrum.flux - num_right
+
+    # Denominator
+    denom = observed_spectrum.uncertainty.array * observed_spectrum.flux.unit
+
+    # Get chi square
+    result = (num/denom)**2
+    chi2 = np.sum(result)
+
+    # Create normalized template spectrum, which will be returned with corresponding chi2
+    normalized_template_spectrum = Spectrum1D(spectral_axis=template_spectrum.spectral_axis,
+                                              flux=template_spectrum.flux*normalization)
+
+    return normalized_template_spectrum, chi2
+
+def template_match(observed_spectrum, spectral_templates):
+    """
+    Find what instance spectral_templates is and run _template_match accordingly. If two template_spectra have the same
```

Comment:
It's confusing for the docstring to reference a private function because the users don't know about the private function.  Maybe re-phrase as something like 
```suggestion
    Find which spectral templates is the best fit to an observed spectrum by computing the chi-squared. If two template_spectra have the same
```